Fix support for mathematical types

tautschnig opened the pull request:
We were missing front-end support for reals and did not consistently support all of integers, naturals, rationals, reals but instead would only handle varying subsets in each place.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Fix support for mathematical types

We were missing front-end support for reals and did not consistently
support all of integers, naturals, rationals, reals but instead would
only handle varying subsets in each place.

polgreen committed:
Fix and demonstrate support for __CPROVER_rational

polgreen commented on src/solvers/flattening/boolbv_width.cpp:
```diff
@@ -206,6 +206,10 @@ const boolbv_widtht::entryt &boolbv_widtht::get_entry(const typet &type) const
   {
     cache_entry = defined_entryt{0};
   }
+  else if(type_id == ID_rational)
```
```suggestion
  else if(type_id == ID_rational || type_id == ID_real || type_id == ID_integer)
```
I added this line because we needed arrays of rationals, but should probably also support arrays of reals and integers.

kroening commented on src/util/real.h:
```diff
@@ -0,0 +1,59 @@
+class realt
+{
+protected:
+  mp_integer integral, fractional;
```
Representing real numbers like floating-point numbers will at the very least irritate. Please add a comment why this is the right choice.

tautschnig replied:
Hmm, which approach would you have chosen instead?

kroening replied:
E.g., you could have represented the algebraic numbers via the root of a polynomial; I'd be curious what the various SMT solvers output as models.

polgreen replied:
z3 and cvc5 seem to output Reals as fractions, unless the denominator is 1, when they just output a singular decimal. 
 
Whilst they use decimals, I've not managed to make it produce a number that isn't an integer as the nominator or denominator, or as a singular decimal.

~~~
(set-logic ALL)
(declare-fun x () Real)
(declare-fun y () Real)
(declare-fun z () Real)
(assert (= y (/ 100 30)))
(assert (= x 0.111111))
(assert (= z 7))
(check-sat)
(get-model)
~~~

z3 says:
~~~
(
  (define-fun z () Real
    7.0)
  (define-fun y () Real
    (/ 10.0 3.0))
  (define-fun x () Real
    (/ 111111.0 1000000.0))
)
~~~

cvc5 says
~~~
(
(define-fun x () Real (/ 111111 1000000))
(define-fun y () Real (/ 10 3))
(define-fun z () Real 7.0)
)
~~~

For irrational numbers, they output the root of a polynomial. 

~~~
(set-logic ALL)
(declare-fun y () Real)
(assert (= (* y y) 2 ))
(check-sat)
(get-model)
~~~

z3 says
~~~
(
  (define-fun y () Real
    (root-obj (+ (^ x 2) (- 2)) 1))
)
~~~

cvc5 segfaults...

kroening replied:
I suspect we'll eventually want the algebraic numbers.

tautschnig replied:
I'll try to address this (even when I'd argue that the current PR already addresses a number of CBMC invariant failures that the included regression tests cause when attempted without the code changes).

kroening replied:
No need to implement this right now; my suggestion is to add a comment, then merge this PR.

tautschnig replied:
I've added a comment and a KNOWNBUG regression test (amounting to the example that @polgreen gave).

tautschnig replied:
I have now added a further commit that removes `realt` and introduces `algebraic_numbert` while also fixing that `KNOWNBUG` test introduced in the prior commit.

tautschnig committed:
Represent (a subset of) real numbers as algebraic numbers

This representation is used by Z3 for it covers a subset of irrational
numbers in addition to rational numbers.

kroening commented on regression/cbmc/real-assignments1/test.desc:
```diff
@@ -0,0 +1,7 @@
+--trace --smt2
+^EXIT=10$
+^SIGNAL=0$
+^  b=100
```
Is this meant to test constant propagation?

tautschnig replied:
This mainly tests that our representation in the trace works.

kroening commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -496,6 +539,37 @@ constant_exprt smt2_convt::parse_literal(
   }
+  else if(
+    src.get_sub().size() == 3 &&
+    src.get_sub()[0].id() == "root-obj") // (root-obj (+ ...) 1)
```
Which solvers emit these? I suppose these are solver-specific, or does the SMT-LIB standard talk about these?

tautschnig replied:
Added a comment.

kroening commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -496,6 +539,37 @@ constant_exprt smt2_convt::parse_literal(
+    src.get_sub().size() == 3 &&
+    src.get_sub()[0].id() == "root-obj") // (root-obj (+ ...) 1)
+  {
+    DATA_INVARIANT_WITH_DIAGNOSTICS(
```
I don't think that a DATA_INVARIANT is the right choice here. This is a property of the output of another tool, which we have very little control over. I'd make this a proper exception.
